### PR TITLE
Change exception types, improve argument checking, and cleanups in mpl_toolkits

### PR DESCRIPTION
**Pull request by oscargus**
## PR Summary

Raise a more exact type.

Better argument checking.

Minor cleanups.

## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [N/A] New features are documented, with examples if plot related.
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Review by anntzer — APPROVED** [review]

File: lib/mpl_toolkits/axes_grid1/axes_divider.py [hunk 1]
```diff
@@ -48,6 +49,7 @@ def __init__(self, fig, pos, horizontal, vertical,
         self._horizontal = horizontal
         self._vertical = vertical
         self._anchor = anchor
```
Comment (QuLogic) [inline lib/mpl_toolkits/axes_grid1/axes_divider.py:51]: Can drop this line now?

File: doc/api/next_api_changes/behavior/23550-OG.rst [hunk 2]
```diff
@@ -0,0 +1,10 @@
+Specified exception types in Grid
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In a few cases an ``Exception`` was thrown when an incorrect argument value
+was set in the `mpl_toolkits.axes_grid1.axes_grid.Grid`
+(= `mpl_toolkits.axisartist.axes_grid.Grid`) constructor. These are replaced
+as follows:
+
+* Providing an incorrect value for ``ngrids`` now raises a ``ValueError``
+* Providing an incorrect type for ``rect`` now raises a ``TypeError``
```
Comment (QuLogic) [inline doc/api/next_api_changes/behavior/23550-OG.rst:10]:
```suggestion
* Providing an incorrect value for *ngrids* now raises a ``ValueError``
* Providing an incorrect type for *rect* now raises a ``TypeError``
```
I _think_ you can also link the errors.

File: lib/mpl_toolkits/axes_grid1/axes_divider.py [hunk 3]
```diff
@@ -512,6 +516,8 @@ def append_axes(self, position, size, pad=None, add_to_figure=True, *,
         **kwargs
             All extra keywords arguments are passed to the created axes.
         """
+        _api.check_in_list(["left", "right", "bottom", "top"],
+                           position=position)
```
Comment (anntzer) [inline lib/mpl_toolkits/axes_grid1/axes_divider.py:520]: FWIW if you're going to change this block I tend to like writing these as
```python
        create_axes, pack_start = _api.check_getitem({
            "left": (self.new_horizontal, True),
            "right": (self.new_horizontal, False),
            "bottom": (self.new_vertical, True),
            "top": (self.new_vertical, False),
        }, position=position)
        ax = create_axes(
            size, pad, pack_start=pack_start, axes_class=axes_class, **kwargs)
```
which makes the exhaustiveness check clearer.

File: lib/mpl_toolkits/axes_grid1/axes_divider.py (hunk 1 above)
Comment (oscargus) [reply lib/mpl_toolkits/axes_grid1/axes_divider.py:51]: Won't some code quality tools complain if it is defined outside of `__init__`? That was the reason anyway.

File: lib/mpl_toolkits/axes_grid1/axes_divider.py (hunk 3 above)
Comment (oscargus) [reply lib/mpl_toolkits/axes_grid1/axes_divider.py:520]: Sure, I didn't really consider the content of the if-clauses, just realized that the check was at the bottom...

**Pull request merged**